src/HermesTransport.InMemory/InMemorySubscription.cs

```diff
@@ -0,0 +1,187 @@
+        
+        // Complete the message queue
+        _messageQueue.Writer.Complete();
+        
+        // Cancel processing
+        _cancellationTokenSource.Cancel();
+
+        if (_processingTask != null)
+        {
+            try
+            {
+                await _processingTask;
+            }
+            catch (OperationCanceledException)
+            {
+                // Expected when cancelling
+            }
+        }
+    }
+
+    public async Task DeliverMessageAsync(IMessage message)
+    {
+        if (!_isActive) return;
+        
+        // Only deliver messages of the correct type
+        if (message is TMessage)
+        {
+            await _messageQueue.Writer.WriteAsync(message);
+        }
+    }
+
+    public void Complete()
+    {
+        _messageQueue.Writer.Complete();
+    }
+
+    private async Task ProcessMessagesAsync(CancellationToken cancellationToken)
+    {
+        try
+        {
+            await foreach (var message in _messageQueue.Reader.ReadAllAsync(cancellationToken))
+            {
+                if (cancellationToken.IsCancellationRequested) break;
+
+                // Cast to the expected type (we already filtered in DeliverMessageAsync)
+                if (message is TMessage typedMessage)
+                {
+                    try
+                    {
+                        if (_options.MaxConcurrency == 1)
+                        {
+                            // Synchronous processing
+                            await ProcessMessageSynchronously(typedMessage, cancellationToken);
+                        }
+                        else
+                        {
+                            // Asynchronous parallel processing
+                            _ = ProcessMessageAsynchronously(typedMessage, cancellationToken);
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        // Log error (in a real implementation, you'd use ILogger)
+                        Console.WriteLine($"Error processing message {message.MessageId}: {ex.Message}");
+                    }
+                }
+            }
+        }
+        catch (OperationCanceledException)
+        {
+            // Expected when stopping
+        }
+    }
+
+    private async Task ProcessMessageSynchronously(TMessage message, CancellationToken cancellationToken)
+    {
+        await _semaphore.WaitAsync(cancellationToken);
+        try
+        {
+            await _handler.HandleAsync(message, cancellationToken);
+        }
+        finally
+        {
+            _semaphore.Release();
+        }
+    }
+
+    private async Task ProcessMessageAsynchronously(TMessage message, CancellationToken cancellationToken)
+    {
+        await _semaphore.WaitAsync(cancellationToken);
+        
+        // Don't use fire-and-forget, await the task completion
+        _ = Task.Run(async () =>
```

Comment:
Don't use Task.Run - why cant we wait for completition of `_handler.HandleAsync` Task?